onnxruntime/core/session/onnxruntime_c_api.cc

```diff
@@ -897,44 +946,106 @@ ORT_API_STATUS_IMPL(OrtApis::GetValue, const OrtValue* value, int index, OrtAllo
 template <typename T>
 static OrtStatus* OrtCreateValueImplSeqHelperMap(const OrtValue* const* in, size_t num_values, OrtValue** out) {
   using SeqType = std::vector<T>;
-  auto vec_ptr = onnxruntime::make_unique<SeqType>();
-  vec_ptr->reserve(num_values);
+  auto seq_ptr = onnxruntime::make_unique<SeqType>();
+  seq_ptr->reserve(num_values);
   for (size_t idx = 0; idx < num_values; ++idx) {
     auto& m = reinterpret_cast<const OrtValue*>(in[idx])->Get<T>();
-    vec_ptr->push_back(m);
+    seq_ptr->push_back(m);
   }
   // create OrtValue with this vector
   auto value = onnxruntime::make_unique<OrtValue>();
-  value->Init(vec_ptr.release(),
+  value->Init(seq_ptr.release(),
               DataTypeImpl::GetType<SeqType>(),
               DataTypeImpl::GetType<SeqType>()->GetDeleteFunc());
   *out = value.release();
   return nullptr;
 }
 
-template <typename T>
-static OrtStatus* OrtCreateValueImplSeqHelper(const OrtValue* const* in, size_t num_values, OrtValue** out) {
-  using SeqType = std::vector<T>;
-  auto vec_ptr = onnxruntime::make_unique<SeqType>();
-  vec_ptr->reserve(num_values);
+template <typename TensorElemType>
+static OrtStatus* OrtCreateValueImplSeqHelperTensor(const Tensor& tensor,
+                                                    Tensor& out) {
+  auto data = tensor.Data<TensorElemType>();
+  if (!data) {
+    return OrtApis::CreateStatus(ORT_FAIL, "Encountered nullptr.");
+  }
+  OrtAllocator* allocator;
+  OrtStatus* st = OrtApis::GetAllocatorWithDefaultOptions(&allocator);
+  if (st) {
+    return st;
+  }
+
+  st = CreateTensorImplForSeq<TensorElemType>(tensor.Shape().GetDims().data(), tensor.Shape().NumDimensions(), out);
+  if (st) {
+    return st;
+  }
+
+  size_t num_elems = tensor.Shape().Size();
+  auto* out_data = out.MutableData<TensorElemType>();
+  for (size_t i = 0; i < num_elems; ++i) {
+    *out_data++ = *data++;
+  }
+  return nullptr;
+}
+
+static OrtStatus* OrtCreateValueImplSeqHelper(const OrtValue* const* in, size_t num_values,
+                                              OrtValue** out) {
+  auto seq_ptr = std::make_unique<TensorSeq>();
+  seq_ptr->tensors.resize(num_values);
+
+  // use the data type of the first tensor as the data type of the seq
+  seq_ptr->dtype = reinterpret_cast<const OrtValue*>(in[0])->Get<Tensor>().DataType();
+
```

Comment:
Should TensorSeq have a ctor that requires the data type to be set?

Reply:
Yeah, let me check if there're other problems in doing this. Will address in follow up PR.